File: lazyllm/tools/rag/doc_node.py
from typing import Optional, Dict, Any, Union, Callable, List
from enum import Enum, auto
from collections import defaultdict
from PIL import Image
from lazyllm import config, reset_on_pickle
from lazyllm.components.utils.file_operate import _image_to_base64
from .global_metadata import RAG_DOC_ID, RAG_DOC_PATH, RAG_KB_ID
import uuid
import threading
import time
import copy

_pickle_blacklist = {'_store', '_node_groups'}


class MetadataMode(str, Enum):
    ALL = auto()
    EMBED = auto()
    LLM = auto()
    NONE = auto()


@reset_on_pickle(('_lock', threading.Lock))
class DocNode:
    def __init__(self, uid: Optional[str] = None, content: Optional[Union[str, List[Any]]] = None,
                 group: Optional[str] = None, embedding: Optional[Dict[str, List[float]]] = None,
                 parent: Optional[Union[str, "DocNode"]] = None, store=None,
                 node_groups: Optional[Dict[str, Dict]] = None, metadata: Optional[Dict[str, Any]] = None,
                 global_metadata: Optional[Dict[str, Any]] = None, *, text: Optional[str] = None):
        if text and content:
            raise ValueError('`text` and `content` cannot be set at the same time.')
        if not content and not text: content = ''
        self._uid: str = uid if uid else str(uuid.uuid4())
        self._content: Optional[Union[str, List[Any]]] = content if content is not None else text
        self._group: Optional[str] = group
        self._embedding: Optional[Dict[str, List[float]]] = embedding or {}
        # metadata: the chunk's meta
        self._metadata: Dict[str, Any] = metadata or {}
        # Global metadata: the file's global metadata (higher level)
        self._global_metadata = global_metadata or {}
        # Metadata keys that are excluded from text for the embed model.
        self._excluded_embed_metadata_keys: List[str] = []
        # Metadata keys that are excluded from text for the LLM.
        self._excluded_llm_metadata_keys: List[str] = []
        # NOTE: node in parent should be id when stored in db (use store to recover): parent: 'uid'
        self._parent: Optional[Union[str, "DocNode"]] = parent
        self._children: Dict[str, List["DocNode"]] = defaultdict(list)
        self._children_loaded = False
        self._store = store
        self._node_groups: Dict[str, Dict] = node_groups or {}
        self._lock = threading.Lock()
        self._embedding_state = set()
        self.relevance_score = None
        self.similarity_score = None

    @property
    def uid(self) -> str:
        return self._uid

    @property
    def group(self) -> str:
        return self._group

    @property
    def text(self) -> str:
        if isinstance(self._content, str):
            return self._content
        elif isinstance(self._content, list):
            if unexcepted := set([type(ele) for ele in self._content if not isinstance(ele, str)]):
                raise TypeError(f"Found non-string element in content: {unexcepted}")
            return '\n'.join(self._content)
        else:
            raise TypeError(f"content type '{type(self._content)}' is neither a str nor a list")

    @property
    def embedding(self):
        return self._embedding

    @embedding.setter
    def embedding(self, v: Optional[Dict[str, List[float]]]):
        self._embedding = v

    def _load_from_store(self, group_name: str, uids: Union[str, List[str]]) -> List["DocNode"]:
        if not self._store or not uids:
            return []
        if isinstance(uids, str):
            uids = [uids]
        nodes = self._store.get_nodes(group_name=group_name, uids=uids,
                                      kb_id=self.global_metadata.get(RAG_KB_ID), display=True)
        for n in nodes:
            n._store = self._store
            n._node_groups = self._node_groups
        return nodes

    @property
    def parent(self) -> Optional["DocNode"]:
        if self._parent and isinstance(self._parent, str) and self._node_groups:
            parent_group = self._node_groups[self._group]["parent"]
            loaded = self._load_from_store(parent_group, self._parent)
            self._parent = loaded[0] if loaded else None
        return self._parent

    @parent.setter
    def parent(self, v: Optional["DocNode"]):
        self._parent = v

    @property
    def children(self) -> Dict[str, List["DocNode"]]:
        if not self._children_loaded and self._store and self._node_groups:
            self._children_loaded = True
            kb_id = self.global_metadata.get(RAG_KB_ID)
            doc_id = self.global_metadata.get(RAG_DOC_ID)
            c_groups = [grp for grp in self._node_groups.keys() if self._node_groups[grp]['parent'] == self._group]
            for grp in c_groups:
                if not self._store.is_group_active(grp):
                    continue
                nodes = self._store.get_nodes(group_name=grp, kb_id=kb_id, doc_ids=[doc_id])
                c_nodes = [n for n in nodes if n._parent in {self, self._uid}]
                self._children[grp] = c_nodes
                for n in self._children[grp]:
                    n._store = self._store
                    n._node_groups = self._node_groups
        return self._children

    @children.setter
    def children(self, v: Dict[str, List["DocNode"]]):
        self._children = v

    @property
    def root_node(self) -> "DocNode":
        node = self
        while isinstance(node._parent, DocNode):
            node = node._parent
        return node

    @property
    def is_root_node(self) -> bool:
        return (not self.parent)

    @property
    def global_metadata(self) -> Dict[str, Any]:
        return self.root_node._global_metadata

    @global_metadata.setter
    def global_metadata(self, global_metadata: Dict) -> None:
        self._global_metadata = global_metadata

    @property
    def metadata(self) -> Dict:
        return self._metadata

    @metadata.setter
    def metadata(self, metadata: Dict) -> None:
        self._metadata = metadata

    @property
    def excluded_embed_metadata_keys(self) -> List:
        return list(set(self.root_node._excluded_embed_metadata_keys + self._excluded_embed_metadata_keys))

    @excluded_embed_metadata_keys.setter
    def excluded_embed_metadata_keys(self, excluded_embed_metadata_keys: List) -> None:
        self._excluded_embed_metadata_keys = excluded_embed_metadata_keys

    @property
    def excluded_llm_metadata_keys(self) -> List:
        return list(set(self.root_node._excluded_llm_metadata_keys + self._excluded_llm_metadata_keys))

    @excluded_llm_metadata_keys.setter
    def excluded_llm_metadata_keys(self, excluded_llm_metadata_keys: List) -> None:
        self._excluded_llm_metadata_keys = excluded_llm_metadata_keys

    @property
    def docpath(self) -> str:
        return self.root_node.global_metadata.get(RAG_DOC_PATH, '')

    @docpath.setter
    def docpath(self, path):
        assert not self.parent, 'Only root node can set docpath'
        self.global_metadata[RAG_DOC_PATH] = str(path)

    def get_children_str(self) -> str:
        return str(
            {key: [node._uid for node in nodes] for key, nodes in self.children.items()}
        )

    def get_parent_id(self) -> str:
        return self.parent._uid if self.parent else ''

    def __str__(self) -> str:
        return (
            f"DocNode(id: {self._uid}, group: {self._group}, content: {self._content}) parent: {self.get_parent_id()}, "
            f"children: {self.get_children_str()}"
        )

    def __repr__(self) -> str:
        return str(self) if config["debug"] else f'<Node id={self._uid}>'

    def __eq__(self, other):
        if isinstance(other, DocNode):
            return self._uid == other._uid
        return False

    def __hash__(self):
        return hash(self._uid)

    def __getstate__(self):
        st = self.__dict__.copy()
        for attr in _pickle_blacklist:
            st[attr] = None
        return st

    def has_missing_embedding(self, embed_keys: Union[str, List[str]]) -> List[str]:
        if isinstance(embed_keys, str): embed_keys = [embed_keys]
        assert len(embed_keys) > 0, "The ebmed_keys to be checked must be passed in."
        if self.embedding is None: return embed_keys
        return [k for k in embed_keys if k not in self.embedding]

    def do_embedding(self, embed: Dict[str, Callable]) -> None:
        generate_embed = {k: e(self.get_text(MetadataMode.EMBED)) for k, e in embed.items()}
        with self._lock:
            self.embedding = self.embedding or {}
            self.embedding = {**self.embedding, **generate_embed}

    def check_embedding_state(self, embed_key: str) -> None:
        while True:
            with self._lock:
                if not self.has_missing_embedding(embed_key):
                    self._embedding_state.discard(embed_key)
                    break
            time.sleep(1)

    def get_content(self) -> str:
        return self.get_text(MetadataMode.LLM)

    def get_metadata_str(self, mode: MetadataMode = MetadataMode.ALL) -> str:
        """Metadata info string."""
        if mode == MetadataMode.NONE:
            return ''

        metadata_keys = set(self.metadata.keys())
        if mode == MetadataMode.LLM:
            for key in self.excluded_llm_metadata_keys:
                if key in metadata_keys:
                    metadata_keys.remove(key)
        elif mode == MetadataMode.EMBED:
            for key in self.excluded_embed_metadata_keys:
                if key in metadata_keys:
                    metadata_keys.remove(key)

        return "\n".join([f"{key}: {self.metadata[key]}" for key in metadata_keys])

    def get_text(self, metadata_mode: MetadataMode = MetadataMode.NONE) -> str:
        metadata_str = self.get_metadata_str(metadata_mode).strip()
        if not metadata_str:
            return self.text if self.text else ''
        return f"{metadata_str}\n\n{self.text}".strip()

    def to_dict(self) -> Dict:
        return dict(content=self._content, embedding=self.embedding, metadata=self.metadata)

    def with_score(self, score):
        node = copy.copy(self)
        node.relevance_score = score
        return node

    def with_sim_score(self, score):
        node = copy.copy(self)
        node.similarity_score = score
        return node


class QADocNode(DocNode):
    def __init__(self, query: str, answer: str, uid: Optional[str] = None, group: Optional[str] = None,
                 embedding: Optional[Dict[str, List[float]]] = None, parent: Optional["DocNode"] = None,
                 metadata: Optional[Dict[str, Any]] = None, global_metadata: Optional[Dict[str, Any]] = None,
                 *, text: Optional[str] = None):
        super().__init__(uid, query, group, embedding, parent, metadata, global_metadata=global_metadata, text=text)
        self._answer = answer.strip()

    @property
    def answer(self) -> str:
        return self._answer

    def get_text(self, metadata_mode: MetadataMode = MetadataMode.NONE) -> str:
        if metadata_mode == MetadataMode.LLM:
            return f'query:\n{self.text}\nanswer\n{self._answer}'
        return super().get_text(metadata_mode)


class ImageDocNode(DocNode):
    def __init__(self, image_path: str, uid: Optional[str] = None, group: Optional[str] = None,
                 embedding: Optional[Dict[str, List[float]]] = None, parent: Optional["DocNode"] = None,
                 metadata: Optional[Dict[str, Any]] = None, global_metadata: Optional[Dict[str, Any]] = None,
                 *, text: Optional[str] = None):
        super().__init__(uid, None, group, embedding, parent, metadata, global_metadata=global_metadata, text=text)
        self._image_path = image_path.strip()
        self._modality = 'image'

    def do_embedding(self, embed: Dict[str, Callable]) -> None:
        for k, e in embed.items():
            emb = e(self.get_content(MetadataMode.EMBED), modality=self._modality)
            generate_embed = {k: emb[0]}

        with self._lock:
            self.embedding = self.embedding or {}
            self.embedding = {**self.embedding, **generate_embed}

    def get_content(self, metadata_mode=MetadataMode.LLM) -> str:
        if metadata_mode == MetadataMode.LLM:
            return Image.open(self._image_path)
        elif metadata_mode == MetadataMode.EMBED:
            image_base64, mime = _image_to_base64(self._image_path)
            return [f"data:{mime};base64,{image_base64}"]
        else:
            return self.get_text()

    @property
    def image_path(self):
        return self._image_path

    def get_text(self) -> str:  # Disable access to self._content
        return self._image_path

    @property
    def text(self) -> str:  # Disable access to self._content
        return self._image_path

File: csrc/src/doc_node.cpp
#include "doc_node.hpp"
#include "utils.hpp"

#include <algorithm>
#include <cctype>
#include <chrono>
#include <cstdint>
#include <iomanip>
#include <sstream>
#include <stdexcept>
#include <thread>

namespace lazyllm {

DocNode::DocNode()
    : _uid(GenerateUUID()),
      _group(),
      _text(),
      _content_is_list(false),
      _parent(nullptr),
      _children_loaded(false),
      _content_hash(),
      _content_hash_dirty(true),
      _relevance_score(0.0),
      _has_relevance_score(false),
      _similarity_score(0.0),
      _has_similarity_score(false) {}

DocNode::DocNode(const std::string& text) : DocNode() {
    set_text(text);
}

DocNode::DocNode(const DocNode& other)
    : _uid(other._uid),
      _group(other._group),
      _text(other._text),
      _content_is_list(other._content_is_list),
      _content_list(other._content_list),
      _embedding(other._embedding),
      _metadata(other._metadata),
      _global_metadata(other._global_metadata),
      _excluded_embed_metadata_keys(other._excluded_embed_metadata_keys),
      _excluded_llm_metadata_keys(other._excluded_llm_metadata_keys),
      _parent(other._parent),
      _children(other._children),
      _children_loaded(other._children_loaded),
      _embedding_state(other._embedding_state),
      _content_hash(other._content_hash),
      _content_hash_dirty(other._content_hash_dirty),
      _relevance_score(other._relevance_score),
      _has_relevance_score(other._has_relevance_score),
      _similarity_score(other._similarity_score),
      _has_similarity_score(other._has_similarity_score) {}

DocNode& DocNode::operator=(const DocNode& other) {
    if (this == &other) {
        return *this;
    }
    _uid = other._uid;
    _group = other._group;
    _text = other._text;
    _content_is_list = other._content_is_list;
    _content_list = other._content_list;
    _embedding = other._embedding;
    _metadata = other._metadata;
    _global_metadata = other._global_metadata;
    _excluded_embed_metadata_keys = other._excluded_embed_metadata_keys;
    _excluded_llm_metadata_keys = other._excluded_llm_metadata_keys;
    _parent = other._parent;
    _children = other._children;
    _children_loaded = other._children_loaded;
    _embedding_state = other._embedding_state;
    _content_hash = other._content_hash;
    _content_hash_dirty = other._content_hash_dirty;
    _relevance_score = other._relevance_score;
    _has_relevance_score = other._has_relevance_score;
    _similarity_score = other._similarity_score;
    _has_similarity_score = other._has_similarity_score;
    return *this;
}

const std::string& DocNode::uid() const {
    return _uid;
}

const std::string& DocNode::group() const {
    return _group;
}

void DocNode::set_group(const std::string& group) {
    _group = group;
}

bool DocNode::content_is_list() const {
    return _content_is_list;
}

const std::vector<std::string>& DocNode::content_list() const {
    return _content_list;
}

const std::string& DocNode::content_text() const {
    return _text;
}

void DocNode::set_content(const std::string& text) {
    set_text(text);
}

void DocNode::set_content(const std::vector<std::string>& lines) {
    _content_is_list = true;
    _content_list = lines;
    _text = JoinLines(lines);
    invalidate_content_hash();
}

void DocNode::set_text(const std::string& text) {
    _text = text;
    _content_is_list = false;
    _content_list.clear();
    invalidate_content_hash();
}

const std::string& DocNode::get_text() const {
    return _text;
}

std::string DocNode::get_text_with_metadata(MetadataMode mode) const {
    const std::string metadata_str = get_metadata_str(mode);
    if (metadata_str.empty()) {
        return _text;
    }
    if (_text.empty()) {
        return metadata_str;
    }
    return metadata_str + "\n\n" + _text;
}

std::string DocNode::content_hash() const {
    if (_content_hash_dirty) {
        _content_hash = Sha256Hex(_text);
        _content_hash_dirty = false;
    }
    return _content_hash;
}

DocNode::EmbeddingVec& DocNode::embedding() {
    return _embedding;
}

const DocNode::EmbeddingVec& DocNode::embedding() const {
    return _embedding;
}

void DocNode::set_embedding(const EmbeddingVec& embed) {
    std::lock_guard<std::mutex> lock(_embedding_mutex);
    _embedding = embed;
}

std::vector<std::string> DocNode::has_missing_embedding(const std::vector<std::string>& embed_keys) const {
    std::vector<std::string> missing;
    if (embed_keys.empty()) {
        return missing;
    }
    std::lock_guard<std::mutex> lock(_embedding_mutex);
    for (const auto& key : embed_keys) {
        if (_embedding.find(key) == _embedding.end()) {
            missing.push_back(key);
        }
    }
    return missing;
}

void DocNode::do_embedding(const std::unordered_map<std::string, EmbeddingFun>& embed) {
    EmbeddingVec generated;
    const std::string input = get_text_with_metadata(MetadataMode::EMBED);
    for (const auto& item : embed) {
        generated[item.first] = item.second(input, "");
    }
    std::lock_guard<std::mutex> lock(_embedding_mutex);
    for (const auto& item : generated) {
        _embedding[item.first] = item.second;
    }
}

void DocNode::set_embedding_value(const std::string& key, const std::vector<float>& value) {
    std::lock_guard<std::mutex> lock(_embedding_mutex);
    _embedding[key] = value;
}

void DocNode::check_embedding_state(const std::string& embed_key) const {
    while (true) {
        {
            std::lock_guard<std::mutex> lock(_embedding_mutex);
            if (_embedding.find(embed_key) != _embedding.end()) {
                _embedding_state.erase(embed_key);
                break;
            }
        }
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }
}

DocNode* DocNode::parent() {
    return _parent;
}

const DocNode* DocNode::parent() const {
    return _parent;
}

void DocNode::set_parent(DocNode* parent) {
    _parent = parent;
}

DocNode::Children& DocNode::children() {
    return _children;
}

const DocNode::Children& DocNode::children() const {
    return _children;
}

void DocNode::set_children(const Children& children) {
    _children = children;
}

DocNode* DocNode::root_node() {
    DocNode* node = this;
    while (node->_parent != nullptr) {
        node = node->_parent;
    }
    return node;
}

const DocNode* DocNode::root_node() const {
    const DocNode* node = this;
    while (node->_parent != nullptr) {
        node = node->_parent;
    }
    return node;
}

bool DocNode::is_root_node() const {
    return _parent == nullptr;
}

DocNode::Metadata& DocNode::metadata() {
    return _metadata;
}

const DocNode::Metadata& DocNode::metadata() const {
    return _metadata;
}

void DocNode::set_metadata(const Metadata& metadata) {
    _metadata = metadata;
}

DocNode::Metadata& DocNode::global_metadata() {
    return root_node()->_global_metadata;
}

const DocNode::Metadata& DocNode::global_metadata() const {
    return root_node()->_global_metadata;
}

void DocNode::set_global_metadata(const Metadata& global_metadata) {
    _global_metadata = global_metadata;
}

std::vector<std::string> DocNode::excluded_embed_metadata_keys() const {
    std::set<std::string> keys;
    const DocNode* root = root_node();
    keys.insert(root->_excluded_embed_metadata_keys.begin(), root->_excluded_embed_metadata_keys.end());
    keys.insert(_excluded_embed_metadata_keys.begin(), _excluded_embed_metadata_keys.end());
    return std::vector<std::string>(keys.begin(), keys.end());
}

void DocNode::set_excluded_embed_metadata_keys(const std::vector<std::string>& keys) {
    _excluded_embed_metadata_keys = keys;
}

std::vector<std::string> DocNode::excluded_llm_metadata_keys() const {
    std::set<std::string> keys;
    const DocNode* root = root_node();
    keys.insert(root->_excluded_llm_metadata_keys.begin(), root->_excluded_llm_metadata_keys.end());
    keys.insert(_excluded_llm_metadata_keys.begin(), _excluded_llm_metadata_keys.end());
    return std::vector<std::string>(keys.begin(), keys.end());
}

void DocNode::set_excluded_llm_metadata_keys(const std::vector<std::string>& keys) {
    _excluded_llm_metadata_keys = keys;
}

std::string DocNode::docpath() const {
    const auto& meta = global_metadata();
    const auto it = meta.find(kRagDocPath);
    if (it == meta.end()) {
        return "";
    }
    return it->second;
}

void DocNode::set_docpath(const std::string& path) {
    if (!is_root_node()) {
        throw std::runtime_error("Only root node can set docpath.");
    }
    global_metadata()[kRagDocPath] = path;
}

std::string DocNode::get_children_str() const {
    std::ostringstream oss;
    oss << "{";
    bool first_group = true;
    for (const auto& item : _children) {
        if (!first_group) {
            oss << ", ";
        }
        first_group = false;
        oss << item.first << ": [";
        bool first_child = true;
        for (const auto* node : item.second) {
            if (!node) {
                continue;
            }
            if (!first_child) {
                oss << ", ";
            }
            first_child = false;
            oss << node->uid();
        }
        oss << "]";
    }
    oss << "}";
    return oss.str();
}

std::string DocNode::get_parent_id() const {
    return _parent ? _parent->uid() : "";
}

std::string DocNode::to_string() const {
    std::ostringstream oss;
    oss << "DocNode(id: " << _uid << ", group: " << _group << ", content: " << _text << ") parent: "
        << get_parent_id() << ", children: " << get_children_str();
    return oss.str();
}

bool DocNode::operator==(const DocNode& other) const {
    return _uid == other._uid;
}

bool DocNode::operator!=(const DocNode& other) const {
    return !(*this == other);
}

std::size_t DocNode::hash() const {
    return std::hash<std::string>()(_uid);
}

std::string DocNode::get_metadata_str(MetadataMode mode) const {
    if (mode == MetadataMode::NONE) {
        return "";
    }
    std::set<std::string> keys;
    for (const auto& item : _metadata) {
        keys.insert(item.first);
    }
    if (mode == MetadataMode::LLM) {
        const auto excluded = excluded_llm_metadata_keys();
        for (const auto& key : excluded) {
            keys.erase(key);
        }
    } else if (mode == MetadataMode::EMBED) {
        const auto excluded = excluded_embed_metadata_keys();
        for (const auto& key : excluded) {
            keys.erase(key);
        }
    }
    std::ostringstream oss;
    bool first = true;
    for (const auto& key : keys) {
        const auto it = _metadata.find(key);
        if (it == _metadata.end()) {
            continue;
        }
        if (!first) {
            oss << "\n";
        }
        first = false;
        oss << key << ": " << it->second;
    }
    return oss.str();
}

std::string DocNode::get_content(MetadataMode mode) const {
    if (mode == MetadataMode::LLM) {
        return get_text_with_metadata(MetadataMode::LLM);
    }
    return get_text_with_metadata(mode);
}

DocNode DocNode::with_score(double score) const {
    DocNode node(*this);
    node._relevance_score = score;
    node._has_relevance_score = true;
    return node;
}

DocNode DocNode::with_sim_score(double score) const {
    DocNode node(*this);
    node._similarity_score = score;
    node._has_similarity_score = true;
    return node;
}

bool DocNode::has_relevance_score() const {
    return _has_relevance_score;
}

bool DocNode::has_similarity_score() const {
    return _has_similarity_score;
}

double DocNode::relevance_score() const {
    return _relevance_score;
}

double DocNode::similarity_score() const {
    return _similarity_score;
}

void DocNode::invalidate_content_hash() {
    _content_hash_dirty = true;
}

QADocNode::QADocNode(const std::string& query, const std::string& answer)
    : DocNode(query), _answer(Trim(answer)) {}

QADocNode::QADocNode(const std::string& query, const std::string& answer, const std::string& uid,
                     const std::string& group)
    : DocNode(query), _answer(Trim(answer)) {
    if (!uid.empty()) {
        _uid = uid;
    }
    _group = group;
}

const std::string& QADocNode::answer() const {
    return _answer;
}

std::string QADocNode::get_text_with_metadata(MetadataMode mode) const {
    if (mode == MetadataMode::LLM) {
        std::ostringstream oss;
        oss << "query:\n" << _text << "\nanswer\n" << _answer;
        return oss.str();
    }
    return DocNode::get_text_with_metadata(mode);
}

ImageDocNode::ImageDocNode(const std::string& image_path)
    : DocNode(image_path), _image_path(Trim(image_path)), _modality("image") {
    set_text(_image_path);
}

ImageDocNode::ImageDocNode(const std::string& image_path, const std::string& uid, const std::string& group)
    : DocNode(image_path), _image_path(Trim(image_path)), _modality("image") {
    if (!uid.empty()) {
        _uid = uid;
    }
    _group = group;
    set_text(_image_path);
}

const std::string& ImageDocNode::image_path() const {
    return _image_path;
}

std::string ImageDocNode::get_content(MetadataMode mode) const {
    if (mode == MetadataMode::EMBED) {
        std::string file_bytes;
        if (!ReadFileBinary(_image_path, &file_bytes)) {
            return "";
        }
        const std::string mime = ImageMimeType(_image_path);
        const std::string base64 = Base64Encode(file_bytes);
        if (mime.empty()) {
            return base64;
        }
        return "data:" + mime + ";base64," + base64;
    }
    return _image_path;
}

void ImageDocNode::do_embedding(const std::unordered_map<std::string, EmbeddingFun>& embed) {
    EmbeddingVec generated;
    const std::string input = get_content(MetadataMode::EMBED);
    for (const auto& item : embed) {
        generated[item.first] = item.second(input, _modality);
    }
    std::lock_guard<std::mutex> lock(_embedding_mutex);
    for (const auto& item : generated) {
        _embedding[item.first] = item.second;
    }
}

std::string ImageDocNode::get_text_with_metadata(MetadataMode mode) const {
    (void)mode;
    return _image_path;
}

} // namespace lazyllm
